### PR TITLE
blockchain: Separate full data context checks.

### PR DESCRIPTION
**Pull request by davecgh**
This separates the block and header context checks to make it more explicit exactly which checks require the availability of the full data for all ancestors and which checks do not.

Currently, all ancestor data is always available due to the way the download and block processing logic is implemented, however, ultimately, the plan is to decouple the chain processing and connection code from the download logic.  In order to help pave the way towards that goal, this more clearly delineates the checks by redefining the current meaning of the concept of contextual checks and creating a new concept of positional checks.

Positional checks are defined as those which depend on the position of a block or header within the block chain and have all of the ancestor block headers available, but do NOT have all of the full block data for
all of the ancestor blocks.

Contextual checks are now defined as those which depend on having the full block data for all of the ancestors available, which also implies all of the ancestor block headers are available too.

To that end, this introduces two new functions named `checkBlockHeaderPositional` and `checkBlockPositional`, moves all of the checks which adhere to the aforementioned semantics to the new
functions, and updates all call sites which invoke the contextual variants to call the positional variants first.

### CODE REVIEW

File: blockchain/accept.go [hunk 1]
```diff
@@ -43,9 +44,17 @@ func (b *BlockChain) maybeAcceptBlock(block *dcrutil.Block, flags BehaviorFlags)
 		return 0, ruleError(ErrInvalidAncestorBlock, str)
 	}
 
-	// The block must pass all of the validation rules which depend on the
-	// position of the block within the block chain.
-	err := b.checkBlockContext(block, prevNode, flags)
+	// The block must pass all of the validation rules which depend on having
+	// the headers of all ancestors availalbe, but do not rely on having the
```
Comment (dnldd) [inline blockchain/accept.go]: s/availalbe/available

File: blockchain/chain.go [hunk 2]
```diff
@@ -1518,11 +1518,22 @@ func (b *BlockChain) forceHeadReorganization(formerBest chainhash.Hash, newBest
...
 		}
 
+		// Perform preliminary sanity checks on the block and its transactions.
 		err = checkBlockSanity(newBestBlock, b.timeSource, BFNone, b.chainParams)
 		if err != nil {
 			return err
 		}
 
+		// The block must pass all of the validation rules which depend on
+		// having the headers of all ancestors availalbe, but do not rely on
```
Comment (dnldd) [inline blockchain/chain.go]: s/availalbe/available

File: blockchain/accept.go (hunk 1 above)
Comment (davecgh) [reply blockchain/accept.go]: Updated.

File: blockchain/chain.go (hunk 2 above)
Comment (davecgh) [reply blockchain/chain.go]: Updated.

**Pull request merged**